### ISSUE TITLE
type inference failure on custom iterable class unpacking

### ISSUE THREAD

**Issue by danmur97** [python/mypy#20435]
**Bug Report**

When unpacking a **custom** iterable class to build another iterable, mypy does not correctly infer the resulting type i.e. `CustomIter[CustomIter[_T]]` is equal to `CustomIter[_T]` (that should not be true)

**To Reproduce**

[proof-of-concept](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict%2Cstrict-bytes%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=715b7ef593ff28fc8a5fb179991ad252)

**Expected Behavior**

Nested iterable unpacking type checking works fine when using standard types (e.g. list[list[_T]]) however custom iterable classes should have the same behavior.

**Actual Behavior**

As presented on the playground proof-of-concept the issue is that mypy **does not** mark the failure on this incorrectly defined function:
```
def case_1(iter_of_iter: CustomIter[list[int]]) -> CustomIter[int]:
    # Mypy fails
    #
    # Mypy does not flag that the return type does not match.
    # It is `CustomIter[list[int]]` instead of `CustomIter[int]`
    return CustomIter.new(*iter_of_iter)
```
**Your Environment**

- details available on the playground example.

**Comment by sterliakov** [comment]
Condensed repro:

```python
# mypy: disable-error-code=empty-body
from __future__ import annotations

from typing import Generic, Iterator, TypeVar

_T = TypeVar("_T")

class CustomIter(Generic[_T]):
    def __iter__(self) -> Iterator[_T]: ...


def case_1(iter_of_iter: CustomIter[list[int]]) -> None:
    reveal_type([*iter_of_iter])  # N: Revealed type is "builtins.list[Any]"
    reveal_type(list(iter_of_iter))  # N: Revealed type is "builtins.list[builtins.list[builtins.int]]"
```

<s>The likely cause is this (very old) mess:

https://github.com/python/mypy/blob/88cf8e08f3e04f797d216869df0f8d5e030aa07c/mypy/checker.py#L1085-L1118

mypy expects a `Generator` for some reason, and `Iterator` is not a `Generator` - the latter has nothing to do with yielded type in general, it can go through the same `__iter__` protocol.</s>

Ough, sorry, that block still looks wrong, but that's irrelevant here. Our problem is here:

https://github.com/python/mypy/blob/9c1406cb8e359082e5e783083c2fa7a68b3f7441/mypy/argmap.py#L201-L207

In this case `is_subtype` should be true (structural subtyping), but `map_instance_to_supertype` can't handle that:

https://github.com/python/mypy/blob/f33c9a3b97f8226eb0156d50be7885ad96815f7c/mypy/maptype.py#L8-L14

**Comment by randolf-scholz** [comment]
This is fixed on my rework star args branch (https://github.com/randolf-scholz/mypy/pull/6) (context: https://github.com/python/mypy/pull/19651#issuecomment-3312947949), not sure when I'll have time to submit / clean up that monster PR though.
